src/pages/workspaces/workspace/notebooks/NotebookLauncher.js

```diff
@@ -32,37 +91,300 @@ const NotebookLauncher = _.flow(
+)(
+  ({ queryParams = {}, notebookName, workspace, workspace: { workspace: { namespace, name }, accessLevel, canCompute }, cluster, refreshClusters, onRequesterPaysError }, ref) => {
+    const [createOpen, setCreateOpen] = useState(false)
+    // Status note: undefined means still loading, null means no cluster
+    const clusterStatus = cluster && cluster.status
+    const [busy, setBusy] = useState()
+    const mode = queryParams['mode']
+
+    return h(Fragment, [
+      (Utils.canWrite(accessLevel) && canCompute && !!mode && clusterStatus === 'Running') ?
+        h(cluster.labels.welderInstallFailed ? WelderDisabledNotebookEditorFrame : NotebookEditorFrame, { key: cluster.clusterName, workspace, cluster, notebookName, mode }) :
+        h(Fragment, [
+          h(PreviewHeader, { queryParams, cluster, refreshClusters, notebookName, workspace, readOnlyAccess: !(Utils.canWrite(accessLevel) && canCompute), onCreateCluster: () => setCreateOpen(true) }),
+          h(NotebookPreviewFrame, { notebookName, workspace })
+        ]),
+      mode && h(ClusterKicker, { cluster, refreshClusters, onNullCluster: () => { setCreateOpen(true) } }),
+      mode && h(ClusterStatusMonitor, { cluster, onClusterStoppedRunning: () => { chooseMode(undefined) } }),
+      h(NewClusterModal, {
+        isOpen: createOpen,
+        namespace, currentCluster: cluster,
+        onDismiss: () => {
+          chooseMode(undefined)
+          setCreateOpen(false)
+        },
+        onSuccess: withErrorReporting('Error creating cluster', async promise => {
+          setCreateOpen(false)
+          setBusy(true)
+          await promise
+          await refreshClusters()
+          setBusy(false)
+        })
+      }),
+      busy && spinnerOverlay
     ])
-})
+  })
 
-const ReadOnlyMessage = ({ notebookName, workspace, workspace: { canCompute, workspace: { namespace, name } } }) => {
-  const [copying, setCopying] = useState(false)
+const FileInUseModal = ({ onDismiss, onCopy, onPlayground, namespace, name, bucketName, lockedBy, canShare }) => {
+  const [lockedByEmail, setLockedByEmail] = useState(null)
 
-  return div({ style: { padding: '1rem 2rem', display: 'flex', alignItems: 'center' } }, [
-    div({ style: { fontSize: 16, fontWeight: 'bold', position: 'absolute' } },
-      ['Viewing read-only']),
-    div({ style: { flexGrow: 1 } }),
-    Utils.cond(
-      [!canCompute, () => h(ButtonOutline, { onClick: () => setCopying(true) }, ['copy to another workspace to edit'])],
-      () => h(ButtonOutline, {
-        href: Nav.getLink('workspace-notebook-launch', { namespace, name, notebookName })
-      }, ['edit in Jupyter'])
+  Utils.useOnMount(() => {
+    const findLockedByEmail = withErrorReporting('Error loading locker information', async () => {
+      const potentialLockers = await findPotentialNotebookLockers(canShare, namespace, name, bucketName)
+      const currentLocker = potentialLockers[lockedBy]
+      setLockedByEmail(currentLocker || null)
```

Comment:
If not present, it'll be undefined:
```suggestion
      setLockedByEmail(currentLocker)
```